In [175]:
!pip install kagglehub

In [176]:
import kagglehub
import pandas as pd

In [177]:
# Download latest version
path = kagglehub.dataset_download("yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring")

In [178]:
print("Path to dataset files:", path)

Path to dataset files: C:\Users\steve\.cache\kagglehub\datasets\yagunnersya\fifa-21-messy-raw-dataset-for-cleaning-exploring\versions\2


In [179]:
# Read the dataset from path
df = pd.read_csv(f"{path}/fifa21 raw data v2.csv")

C:\Users\steve\AppData\Local\Temp\ipykernel_23608\1496996970.py:2: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"{path}/fifa21 raw data v2.csv")


In [180]:
df

ID               Name                      LongName  \
0      158023           L. Messi                  Lionel Messi   
1       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389           J. Oblak                     Jan Oblak   
3      192985       K. De Bruyne               Kevin De Bruyne   
4      190871          Neymar Jr    Neymar da Silva Santos Jr.   
...       ...                ...                           ...   
18974  247223             Xia Ao                        Ao Xia   
18975  258760           B. Hough                     Ben Hough   
18976  252757        R. McKinley                Ronan McKinley   
18977  243790       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520          Zhou Xiao                     Xiao Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/players/200/389/21_60.png   
3      https://cdn.sofifa.com/players/192/985/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
...                                                 ...   
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   
...                                                  ...         ...  ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                         Club  ...     A/W     D/W   IR PAC  \
0        93   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85   
1        92   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89   
2        91   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87   
3        91   91      \n\n\n\nManchester City  ...    High    High  4 ★  76   
4        91   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91   
...     ...  ...                          ...  ...     ...     ...  ...  ..   
18974    47   55           \n\n\n\nWuhan Zall  ...  Medium  Medium  1 ★  64   
18975    47   67      \n\n\n\nOldham Athletic  ...  Medium  Medium  1 ★  64   
18976    47   65           \n\n\n\nDerry City  ...  Medium  Medium  1 ★  63   
18977    47   57     \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  58   
18978    47   57     \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  62   

      SHO  PAS DRI DEF PHY Hits  
0      92   91  95  38  65  771  
1      93   81  89  35  77  562  
2      92   78  90  52  90  150  
3      86   93  88  64  78  207  
4      85   86  94  36  59  595  
...    ..  ...  ..  ..  ..  ...  
18974  28   26  38  48  51  NaN  
18975  40   48  49  35  45  NaN  
18976  39   44  46  40  53  NaN  
18977  49   41  49  30  44  NaN  
18978  22   39  42  45  55  NaN  

[18979 rows x 77 columns]

18979 rows and 77 columns! Let's clean it!<br>
And let's first drop duplicates and check for null values

In [182]:
# df = df.drop_duplicates()
#sorted(df.isna().sum(), reverse=True)
missing_values = df.isna().sum().nlargest(5).sort_values(ascending=False).reset_index()
missing_values.columns = ['Column', 'Missing values']
missing_values

Column  Missing values
0  Loan Date End           17966
1           Hits            2595
2             ID               0
3           Name               0
4       LongName               0

We're eventually gonna have to do something about these 2595 missing values in the Hits column and decide on what to do with the Loan Date End column. <br>
Let's continue with looking at all the columns to see if there are any that we can straight up remove

In [184]:
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

In [185]:
df['BOV']

0        93
1        92
2        91
3        91
4        91
         ..
18974    49
18975    51
18976    49
18977    48
18978    50
Name: BOV, Length: 18979, dtype: int64

The only column I'm going to drop is the Name column and split the LongName column into First and LastName.<br> And we might get rid of is the Hits column as well

In [187]:
# df = df.drop(columns=['Name']) We are not dropping the Name column yet haha!

In [188]:
df['LongName']

0                        Lionel Messi
1        C. Ronaldo dos Santos Aveiro
2                           Jan Oblak
3                     Kevin De Bruyne
4          Neymar da Silva Santos Jr.
                     ...             
18974                          Ao Xia
18975                       Ben Hough
18976                  Ronan McKinley
18977                    Zhen'ao Wang
18978                       Xiao Zhou
Name: LongName, Length: 18979, dtype: object

Okay, wow, it's not as easy as dropping the Name column and splitting the name in LongName which we can see with Mr. Portugal haha!<br>
This is gonna require some finessing. I want to believe that the only exception is the ones that Ronaldo has; that the first name is simply a letter. And in these cases, we need to grab the first name from the name column. And after this operation, *then* we can drop the Name column

In [190]:
exceptions = df[df['LongName'].apply(lambda x: x[1] == '.' if len(x) > 1 else False)]
# exceptions = df[df['LongName'].str.contains(r'\.')] This would check simply for the presence of a dot but that doesn't give us what we want
exceptions

# Saving the IDs of the exceptions to double-check after the transformation of df
exceptions_ids = list(exceptions['ID'])

These are the exceptions we need to handle. In all cases except for player #256612 and #190883 we can simply grab the first name from the Name column like we initially planned. For these other two, there will be two cases; either Name.split() is of length 1 (which is the case for Evanilson), or there is no first name to be grabbed (which is the case for A. DeLaGarza). Let's start coding this up:

In [193]:
for index, row in exceptions.iterrows():
    if len(row['Name'].split()) > 1 and row['Name'][1] != '.':
        # print(f"Name: {row['Name']}, LongName: {row['LongName']}")
        first_name = row['Name'].split()[0]
        df.at[index, 'LongName'].replace('.', '')
        df.at[index, 'LongName'] = f"{first_name} {row['LongName'][2:]}"

There is essentially an else: pass in the for loop above since in the two exceptions I think the best solution is to just leave it as is (F. Evanilson de Lima Barbosa and A.J. DeLaGarza)

In [195]:
df_exceptions_checked = df[df['ID'].isin(exceptions_ids)]
df_exceptions_checked

ID                 Name                              LongName  \
1      20801    Cristiano Ronaldo  Cristiano  Ronaldo dos Santos Aveiro   
729   230406        Guilson Paiva           Guilson  Pierre Paiva Souza   
733   230212  Antônio Chiamuloira    Antônio  Benjamin Chiamuloira Paes   
2764  256612            Evanilson          F. Evanilson de Lima Barbosa   
3224  230407     Miltisco Paulino    Miltisco  Vanderlei Paulino Kenedy   
8285  230365       Sérgio Espinho       Sérgio  Jefferson Espinho Sobri   
9156  190883         A. DeLaGarza                        A.J. DeLaGarza   

                                              photoUrl  \
1     https://cdn.sofifa.com/players/020/801/21_60.png   
729   https://cdn.sofifa.com/players/230/406/21_60.png   
733   https://cdn.sofifa.com/players/230/212/21_60.png   
2764  https://cdn.sofifa.com/players/256/612/21_60.png   
3224  https://cdn.sofifa.com/players/230/407/21_60.png   
8285  https://cdn.sofifa.com/players/230/365/21_60.png   
9156  https://cdn.sofifa.com/players/190/883/21_60.png   

                                              playerUrl Nationality  Age  \
1     http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
729   http://sofifa.com/player/230406/g-pierre-paiva...      Brazil   24   
733   http://sofifa.com/player/230212/a-benjamin-chi...      Brazil   20   
2764  http://sofifa.com/player/256612/f-evanilson-de...      Brazil   20   
3224  http://sofifa.com/player/230407/m-vanderlei-pa...      Brazil   20   
8285  http://sofifa.com/player/230365/s-jefferson-es...      Brazil   24   
9156  http://sofifa.com/player/190883/aj-delagarza/2...        Guam   32   

      ↓OVA  POT                   Club  ...     A/W     D/W   IR PAC SHO  PAS  \
1       92   92       \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
729     78   78       \n\n\n\nFlamengo  ...  Medium  Medium  1 ★  83  69   75   
733     78   78  \n\n\n\nInternacional  ...  Medium  Medium  1 ★  68  48   60   
2764    73   87       \n\n\n\nFC Porto  ...  Medium     Low  1 ★  72  76   54   
3224    72   72         \n\n\n\nSantos  ...  Medium  Medium  1 ★  62  46   57   
8285    67   67  \n\n\n\nVasco da Gama  ...  Medium  Medium  1 ★  58  56   63   
9156    66   66    \n\n\n\nInter Miami  ...  Medium    High  1 ★  67  48   61   

     DRI DEF PHY Hits  
1     89  35  77  562  
729   79  42  68    7  
733   52  79  75   16  
2764  70  25  58  655  
3224  56  72  68    2  
8285  68  64  72    1  
9156  62  63  68   10  

[7 rows x 77 columns]

In [196]:
# *Now* we can drop the Name column
#df = df.drop(columns=['Name'])
#df

In [197]:
# Instead of dropping the Name column, we can simply remove it to FirstName
df = df.rename(columns={'Name':'FirstName'})

# And now on to splitting the LongName column
df['FirstName'] = df['LongName'].apply(lambda x: x.split()[0])
df['LongName'] = df['LongName'].apply(lambda x: " ".join(x.split()[1:]))  # Remove the first name in LongName

df = df.rename(columns={'LongName':'LastName'})

In [215]:
df

ID  FirstName                   LastName  \
0      158023     Lionel                      Messi   
1       20801  Cristiano  Ronaldo dos Santos Aveiro   
2      200389        Jan                      Oblak   
3      192985      Kevin                  De Bruyne   
4      190871     Neymar        da Silva Santos Jr.   
...       ...        ...                        ...   
18974  247223         Ao                        Xia   
18975  258760        Ben                      Hough   
18976  252757      Ronan                   McKinley   
18977  243790    Zhen'ao                       Wang   
18978  252520       Xiao                       Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/players/200/389/21_60.png   
3      https://cdn.sofifa.com/players/192/985/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
...                                                 ...   
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   
...                                                  ...         ...  ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                         Club  ...     A/W     D/W   IR PAC  \
0        93   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85   
1        92   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89   
2        91   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87   
3        91   91      \n\n\n\nManchester City  ...    High    High  4 ★  76   
4        91   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91   
...     ...  ...                          ...  ...     ...     ...  ...  ..   
18974    47   55           \n\n\n\nWuhan Zall  ...  Medium  Medium  1 ★  64   
18975    47   67      \n\n\n\nOldham Athletic  ...  Medium  Medium  1 ★  64   
18976    47   65           \n\n\n\nDerry City  ...  Medium  Medium  1 ★  63   
18977    47   57     \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  58   
18978    47   57     \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  62   

      SHO  PAS DRI DEF PHY Hits  
0      92   91  95  38  65  771  
1      93   81  89  35  77  562  
2      92   78  90  52  90  150  
3      86   93  88  64  78  207  
4      85   86  94  36  59  595  
...    ..  ...  ..  ..  ..  ...  
18974  28   26  38  48  51  NaN  
18975  40   48  49  35  45  NaN  
18976  39   44  46  40  53  NaN  
18977  49   41  49  30  44  NaN  
18978  22   39  42  45  55  NaN  

[18979 rows x 77 columns]

To re-cap:
1. We found all the abbreviated name exceptions
2. We changed the first name in LongName wherever possible
3. Finally, we changed the name of the Name column to FirstName, took it and removed it from LongName, and changed the name of this column to LastName

I have optimized this in the 'optimized cleaning' notebook.<br><br>

In [246]:
df.columns

Index(['ID', 'FirstName', 'LastName', 'photoUrl', 'playerUrl', 'Nationality',
       'Age', '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height',
       'Weight', 'Preferred Foot', 'BOV', 'Best Position', 'Joined',
       'Loan Date End', 'Value', 'Wage', 'Release Clause', 'Attacking',
       'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys',
       'Skill', 'Dribbling', 'Curve', 'FK Accuracy', 'Long Passing',
       'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility',
       'Reactions', 'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina',
       'Strength', 'Long Shots', 'Mentality', 'Aggression', 'Interceptions',
       'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending',
       'Marking', 'Standing Tackle', 'Sliding Tackle', 'Goalkeeping',
       'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning',
       'GK Reflexes', 'Total Stats', 'Base Stats', 'W/F', 'SM', 'A/W', 'D/W',
       'IR', 'PAC', 'SHO', 'PAS', 'DRI', 'D

On to Nationality!

In [236]:
sorted(df['Nationality'].unique())

['Afghanistan',
 'Albania',
 'Algeria',
 'Andorra',
 'Angola',
 'Antigua & Barbuda',
 'Argentina',
 'Armenia',
 'Aruba',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Barbados',
 'Belarus',
 'Belgium',
 'Belize',
 'Benin',
 'Bermuda',
 'Bolivia',
 'Bosnia Herzegovina',
 'Brazil',
 'Bulgaria',
 'Burkina Faso',
 'Burundi',
 'Cameroon',
 'Canada',
 'Cape Verde',
 'Central African Republic',
 'Chad',
 'Chile',
 'China PR',
 'Chinese Taipei',
 'Colombia',
 'Comoros',
 'Congo',
 'Costa Rica',
 'Croatia',
 'Cuba',
 'Curacao',
 'Cyprus',
 'Czech Republic',
 'DR Congo',
 'Denmark',
 'Dominican Republic',
 'Ecuador',
 'Egypt',
 'El Salvador',
 'England',
 'Equatorial Guinea',
 'Eritrea',
 'Estonia',
 'Ethiopia',
 'Faroe Islands',
 'Finland',
 'France',
 'Gabon',
 'Gambia',
 'Georgia',
 'Germany',
 'Ghana',
 'Greece',
 'Grenada',
 'Guam',
 'Guatemala',
 'Guinea',
 'Guinea Bissau',
 'Guyana',
 'Haiti',
 'Honduras',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Iran',
 'Iraq',
 'Isra

Looking at this, there doesn't seem to be any cleaning necessary in the Natinality column!

In [253]:
df['Club'].value_counts()

Club
No Club                           237
\n\n\n\nFC Barcelona               33
\n\n\n\nBurnley                    33
\n\n\n\nRC Celta                   33
\n\n\n\nTorino                     33
                                 ... 
\n\n\n\nMelbourne Victory          18
\n\n\n\nBrisbane Roar              18
\n\n\n\nAdelaide United            18
\n\n\n\nPerth Glory                18
\n\n\n\nCentral Coast Mariners     18
Name: count, Length: 682, dtype: int64

For the Club column, we definitely want to remove the '\n' characters. We can do this pretty simply with replace

In [255]:
df['Club'] = df['Club'].str.replace('\n', '')

In [257]:
df

ID  FirstName                   LastName  \
0      158023     Lionel                      Messi   
1       20801  Cristiano  Ronaldo dos Santos Aveiro   
2      200389        Jan                      Oblak   
3      192985      Kevin                  De Bruyne   
4      190871     Neymar        da Silva Santos Jr.   
...       ...        ...                        ...   
18974  247223         Ao                        Xia   
18975  258760        Ben                      Hough   
18976  252757      Ronan                   McKinley   
18977  243790    Zhen'ao                       Wang   
18978  252520       Xiao                       Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/players/200/389/21_60.png   
3      https://cdn.sofifa.com/players/192/985/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
...                                                 ...   
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   
...                                                  ...         ...  ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                 Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0        93   93         FC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1        92   92             Juventus  ...    High     Low  5 ★  89  93   81   
2        91   93      Atlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3        91   91      Manchester City  ...    High    High  4 ★  76  86   93   
4        91   91  Paris Saint-Germain  ...    High  Medium  5 ★  91  85   86   
...     ...  ...                  ...  ...     ...     ...  ...  ..  ..  ...   
18974    47   55           Wuhan Zall  ...  Medium  Medium  1 ★  64  28   26   
18975    47   67      Oldham Athletic  ...  Medium  Medium  1 ★  64  40   48   
18976    47   65           Derry City  ...  Medium  Medium  1 ★  63  39   44   
18977    47   57     Dalian YiFang FC  ...  Medium  Medium  1 ★  58  49   41   
18978    47   57     Dalian YiFang FC  ...  Medium  Medium  1 ★  62  22   39   

      DRI DEF PHY Hits  
0      95  38  65  771  
1      89  35  77  562  
2      90  52  90  150  
3      88  64  78  207  
4      94  36  59  595  
...    ..  ..  ..  ...  
18974  38  48  51  NaN  
18975  49  35  45  NaN  
18976  46  40  53  NaN  
18977  49  30  44  NaN  
18978  42  45  55  NaN  

[18979 rows x 77 columns]

Let's look for other anomalies and oddities in the Club column systematically

In [274]:
# 1. Check for Unexpected Characters or Patterns
unusual_clubs = df[df['Club'].str.contains(r'[^a-zA-Z0-9 ]', na=False)]['Club'].unique()
unusual_clubs

array(['Atlético Madrid', 'Paris Saint-Germain', 'FC Bayern München',
       'Borussia Mönchengladbach', 'Grêmio', 'Medipol Başakşehir FK',
       'Deportivo Alavés', 'Atlético Mineiro', 'São Paulo',
       '1. FC Union Berlin', 'AS Saint-Étienne', 'Tigres U.A.N.L.',
       'Beşiktaş JK', 'Brighton & Hove Albion', 'Fenerbahçe SK',
       'Guangzhou R&F FC', 'Club América', 'Vitória Guimarães',
       'Club León', 'Club Atlético Lanús', 'CD Leganés',
       'Vélez Sarsfield', '1. FC Köln', 'Shijiazhuang Ever Bright F.C.',
       'Fatih Karagümrük S.K.', 'Universidad Católica', 'Goiás',
       'Qingdao Huanghai F.C.', 'Olimpia Asunción', 'Çaykur Rizespor',
       'Cádiz CF', '1. FSV Mainz 05', 'UD Almería', 'FC København',
       'Nîmes Olympique', 'Colo-Colo', 'Malmö FF',
       'Atlético Clube Goianiense', 'Fortuna Düsseldorf', 'Famalicão',
       'Peñarol', 'U.N.A.M.', "Newell's Old Boys", 'Atlético de San Luis',
       'Club Atlético Colón', 'Clube Sport Marítimo', 'MKE Ankaragücü',
